File: FBDeviceControl/Commands/FBDeviceLocationCommands.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBDeviceLocationCommands.h"

#import "FBDevice.h"
#import "FBDevice+Private.h"
#import "FBDeviceControlError.h"
#import "FBAMDServiceConnection.h"
#import "FBAFCConnection.h"

@interface FBDeviceLocationCommands ()

@property (nonatomic, weak, readonly) FBDevice *device;

@end

@implementation FBDeviceLocationCommands

#pragma mark Initializers

+ (instancetype)commandsWithTarget:(FBDevice *)target
{
  return [[self alloc] initWithDevice:target];
}

- (instancetype)initWithDevice:(FBDevice *)device
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _device = device;

  return self;
}

#pragma mark FBDeviceLocationCommands

static const int StartCommand = 0x0000000;

- (FBFuture<NSNull *> *)overrideLocationWithLongitude:(double)longitude latitude:(double)latitude
{
  return [[[self.device
    mountDeveloperDiskImage]
    onQueue:self.device.workQueue pushTeardown:^(id _) {
      return [self.device startService:@"com.apple.dt.simulatelocation"];
    }]
    onQueue:self.device.workQueue pop:^ FBFuture<NSNull *> * (FBAMDServiceConnection *connection) {
      NSData *start = [[NSData alloc] initWithBytes:&StartCommand length:sizeof(StartCommand)];
      NSError *error = nil;
      id<FBAMDServiceConnectionTransfer> transfer = connection.serviceConnectionWrapped;
      if (![transfer send:start error:&error]) {
        return [FBFuture futureWithError:error];
      }
      NSString *value = [NSString stringWithFormat:@"%f", latitude];
      NSData *data = [[NSData alloc] initWithBytes:value.UTF8String length:strlen(value.UTF8String)];
      if (![transfer sendWithLengthHeader:data error:&error]) {
        return [FBFuture futureWithError:error];
      }
      value = [NSString stringWithFormat:@"%f", longitude];
      data = [[NSData alloc] initWithBytes:value.UTF8String length:strlen(value.UTF8String)];
      if (![transfer sendWithLengthHeader:data error:&error]) {
        return [FBFuture futureWithError:error];
      }
      return FBFuture.empty;
    }];
}

@end

File: FBDeviceControl/Commands/FBDeviceLocationCommands.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@preconcurrency import FBControlCore
import Foundation

// swiftlint:disable force_unwrapping

private let StartCommand: UInt32 = 0x00000000

@objc(FBDeviceLocationCommands)
public class FBDeviceLocationCommands: NSObject, FBiOSTargetCommand {
  private weak var device: FBDevice?

  // MARK: - Initializers

  public class func commands(with target: any FBiOSTarget) -> Self {
    return self.init(device: target as! FBDevice)
  }

  required init(device: FBDevice) {
    self.device = device
    super.init()
  }

  // MARK: - FBLocationCommands (legacy FBFuture entry point)

  public func overrideLocation(withLongitude longitude: Double, latitude: Double) -> FBFuture<NSNull> {
    fbFutureFromAsync { [self] in
      try await overrideLocationAsync(withLongitude: longitude, latitude: latitude)
      return NSNull()
    }
  }

  // MARK: - Async

  fileprivate func overrideLocationAsync(withLongitude longitude: Double, latitude: Double) async throws {
    guard let device else {
      throw FBDeviceControlError().describe("Device is nil").build()
    }
    _ = try await bridgeFBFuture(device.ensureDeveloperDiskImageIsMounted())
    try await withFBFutureContext(device.startService("com.apple.dt.simulatelocation")) { connection in
      var start = StartCommand
      let startData = Data(bytes: &start, count: MemoryLayout<UInt32>.size)
      try connection.send(startData)

      let latitudeString = "\(latitude)"
      let latitudeData = latitudeString.data(using: .utf8)!
      try connection.send(withLengthHeader: latitudeData)

      let longitudeString = "\(longitude)"
      let longitudeData = longitudeString.data(using: .utf8)!
      try connection.send(withLengthHeader: longitudeData)
    }
  }
}

// MARK: - FBDevice+AsyncLocationCommands

extension FBDevice: AsyncLocationCommands {

  public func overrideLocation(longitude: Double, latitude: Double) async throws {
    try await locationCommands().overrideLocationAsync(withLongitude: longitude, latitude: latitude)
  }
}

// MARK: - FBDevice+FBLocationCommands

extension FBDevice: FBLocationCommands {

  @objc(overrideLocationWithLongitude:latitude:)
  public func overrideLocation(withLongitude longitude: Double, latitude: Double) -> FBFuture<NSNull> {
    do {
      return try locationCommands().overrideLocation(withLongitude: longitude, latitude: latitude)
    } catch {
      return FBFuture(error: error)
    }
  }
}
